### ISSUE TITLE
Incorrect parsing of for loop

### ISSUE THREAD

**Issue by AllegroBarry** [koalaman/shellcheck#2021]
#### For bugs
- Rule Id : SC1073, SC1061, SC1036, SC1062, SC1072
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

declare -A MyList

for X in "[A]=1 [B]=2 [C]=3" ; do
  eval MyList+=($X)
  done

```

#### Here's what shellcheck currently says:
```
Line 5	SC1073: Couldn't parse this for loop. Fix to allow more checks.
Line 5	SC1061: Couldn't find 'done' for this 'do'.
Line 6	SC1036: '(' is invalid here. Did you forget to escape it?
Line 6	SC1062: Expected 'done' matching previously mentioned 'do'.
Line 6	SC1072: Expected 'done'. Fix any mentioned problems and try again.
```
#### Here's what I wanted or expected to see:

NOTE: This /bin/bash code is absolutely correct, and executes just fine in the real world. Perhaps your syntax checker doesn't understand associative arrays?:

#!/bin/bash

declare -A DownloadList

for X in "[A]=1 [B]=2 [C]=3" ; do
  eval DownloadList+=($X)
  done

Somehow your for-loop parser is failing to understand this simple loop.

Good luck!

**Comment by AllegroBarry** [comment]
I also tried

`for X in [A]=1 [B]=2 [C]=3 ; do`

and

`for X in '[A]=1'  '[B]=2'  '[C]=3'  ; do`

Nothing works. (That is to say, the ShellCheck tool flags all as errors when they are not. But in the real world, in /bin/bash, they all work fine.)

Further (and I should have done this in the first place), if I pass all three variants to ShellCheck at once, as in...

```bash
#!/bin/bash

declare -A MyList1

for X in "[A]=1 [B]=2 [C]=3" ; do
  eval MyList1+=($X)
  done


declare -A MyList2

for X in [A]=1 [B]=2 [C]=3 ; do
  eval MyList2+=($X)
  done


declare -A MyList3

for X in '[A]=1' '[B]=2' '[C]=3' ; do
  eval MyList3+=($X)
  done
```
...then an error is reported only for the first for-loop, then ShellCheck seems to stop, or crash, and doesn't continue checking the rest of the script.

I don't believe the problem is with the various forms of the for-loop itself, but the fact that we later attempt to reference the X variable in an associative array context.

If I remove all instances of associative arrays, and change `eval...` to a simple `echo`, then ShellCheck has no trouble. In other words, this works perfectly, and reports no syntax issues (Ok, one small issue):

```bash
#!/bin/bash

for X in "[A]=1 [B]=2 [C]=3" ; do
  echo "$X"
  done


for X in [A]=1 [B]=2 [C]=3 ; do
  echo "$X"
  done


for X in '[A]=1' '[B]=2' '[C]=3' ; do
  echo "$X"
  done
```

**Comment by koalaman** [comment]
This is unrelated to loops and associative arrays. It happens when `eval`ing a string without escaping the parentheses. You can see this described on the wiki page for the parenthesis error: https://github.com/koalaman/shellcheck/wiki/SC1036

As described, you can work around it by quoting the eval'd string, e.g. `eval "MyList2+=($X)"`. This does not change the behavior in bash.

The bash parser has some odd quirks due to trying to allow this. For example, you can do `name=foo; eval $name=val` and `eval foo=(42)`, but you can't do `eval $name=(42)`. Similarly, `export() { echo "$1"; }; export foo=(bar)` is syntactically valid (though you can't guess what it does), while `othername() { echo $1; }; othername foo=(bar)` is invalid. The same behavior is why `export var=$foo` does not require quotes in bash, but does on other shells.

ShellCheck copies many of these, but unfortunately trying to mimic them all tends to lead to more unhelpful messages in less advanced cases. Obviously it's never great to have to change working code to please some other tool, but hopefully it's an acceptable price for offering better pinpointed errors to users who are less familiar with the shell.

**Comment by AllegroBarry** [comment]
Thanks very much for the detailed reply. I really appreciate it. Unfortunately, you closed the bug report so I can't 
post any more comments there. But I believe some comments are still warranted.

So the real problem with my script, as you explain, is this line:

   eval MyList+=($X)

I accept that, and I acknowledge that in order to satisfy shellcheck I should quote the assignment, or make some other 
geeky change. Fine.

However, I would gently and humbly submit to you that shellcheck still has a problem; namely, that it reports these issues:

   Line 5	SC1073: Couldn't parse this for loop. Fix to allow more checks.
   Line 5	SC1061: Couldn't find 'done' for this 'do'.

Those are extremely misleading messages. You said yourself that this is unrelated to for-loops. I spent a fair amount of 
time trying to figure out what was wrong with my for-loop, when there never was anything wrong with it. And, guess 
what... the "done" is right there, it's not missing. So why couldn't shellcheck find it? bash finds it just fine.  I 
would expect for a tool like shellcheck to find actual problems in scripts, not invent problems that don't exist.

Also, you apparently overlooked another problem that I mentioned further down in my submission. I created a script with 
three different variants of the for-loop (thinking, due to shellcheck's erroneous error report, that there was a problem 
with the for-loop). But shellcheck reports an issue with the first loop then stops, and doesn't evaluate the rest of the 
script. So one might be lured into believing that there is a problem only with the first loop, and that the rest of the 
script is Ok. But we know that the exact same problem should be found with all three loops. Why does shellcheck stop?

Barry






On 7/25/20 15:03, Vidar Holen wrote:> This is unrelated to loops and associative arrays. It happens when `eval`ing a 
string without escaping the parentheses. You can see this described on the wiki page for the parenthesis error: 
https://github.com/koalaman/shellcheck/wiki/SC1036
 >
 > As described, you can work around it by quoting the eval'd string, e.g. `eval "MyList2+=($X)"`. This does not change 
the behavior in bash.
 >
 > The bash parser has some odd quirks due to trying to allow this. For example, you can do `name=foo; eval $name=val` 
and `eval foo=(42)`, but you can't do `eval $name=(42)`. Similarly, `export() { echo "$1"; }; export foo=(bar)` is 
syntactically valid (though you can't guess what it does), while `othername() { echo $1; }; othername foo=(bar)` is 
invalid. The same behavior is why `export var=$foo` does not require quotes in bash, but does on other shells.
 >
 > ShellCheck copies many of these, but unfortunately trying to mimic them all tends to lead to more unhelpful messages 
in less advanced cases. Obviously it's never great to have to change working code to please some other tool, but 
hopefully it's an acceptable price for offering better pinpointed errors to users who are less familiar with the shell.
 >
